x-pack/plugins/threat_intelligence/public/modules/timeline/hooks/use_investigate_in_timeline.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useContext } from 'react';
+import moment from 'moment';
+import { DataProvider } from '@kbn/timelines-plugin/common';
+import { generateDataProvider } from '../lib/data_provider';
+import { SecuritySolutionContext } from '../../../containers/security_solution_context';
+import { getIndicatorFieldAndValue } from '../../indicators/lib/field_value';
+import { unwrapValue } from '../../indicators/lib/unwrap_value';
+import { EMPTY_VALUE } from '../../../../common/constants';
+import {
+  Indicator,
+  IndicatorFieldEventEnrichmentMap,
+  RawIndicatorFieldId,
+} from '../../../../common/types/indicator';
+
+export interface UseInvestigateInTimelineParam {
+  /**
+   * Indicator used to retrieve the field and value then passed to the Investigate in Timeline logic
+   */
+  indicator: Indicator;
+}
+
+export interface UseInvestigateInTimelineValue {
+  onClick: (() => Promise<void>) | undefined;
+}
+
+/**
+ * Custom hook that gets an {@link Indicator}, retrieves the field (from the RawIndicatorFieldId.Name)
+ * and value, then creates DataProviders used to do the Investigate in Timeline logic
+ * (see /kibana/x-pack/plugins/security_solution/public/threat_intelligence/use_investigate_in_timeline.ts)
+ */
+export const useInvestigateInTimeline = ({
+  indicator,
+}: UseInvestigateInTimelineParam): UseInvestigateInTimelineValue => {
+  const securitySolutionContext = useContext(SecuritySolutionContext);
+
+  const { key, value } = getIndicatorFieldAndValue(indicator, RawIndicatorFieldId.Name);
+
+  if (!value || value === EMPTY_VALUE || !key) {
```

Comment:
we repeat this peace of logic in many places, I think we should at least move it to a utility function, ideally think about a model of our Indicator and about methods of this model. The best illustration of why it's bad to repeat business logic even if it's super simple is in `indicator_value_actions.tsx` where we have `!key || value === EMPTY_VALUE || !key` which is incorrect

Reply:
yeah I extracted it, looks much better to me that way, thanks!